source/egg/math/Matrix.cc

```diff
@@ -24,28 +26,228 @@ Matrix34f::Matrix34f(f32 _e00, f32 _e01, f32 _e02, f32 _e03, f32 _e10, f32 _e11,
+    f32 yy = 2.0f * q.v.y * q.v.y;
+    f32 zz = 2.0f * q.v.z * q.v.z;
+    f32 xx = 2.0f * q.v.x * q.v.x;
+    f32 xy = 2.0f * q.v.x * q.v.y;
+    f32 xz = 2.0f * q.v.x * q.v.z;
+    f32 yz = 2.0f * q.v.y * q.v.z;
+    f32 wz = 2.0f * q.w * q.v.z;
+    f32 wx = 2.0f * q.w * q.v.x;
+    f32 wy = 2.0f * q.w * q.v.y;
+
+    mtx[0][0] = 1.0f - yy - zz;
+    mtx[0][1] = xy - wz;
+    mtx[0][2] = xz + wy;
+
+    mtx[1][0] = xy + wz;
+    mtx[1][1] = 1.0f - xx - zz;
+    mtx[1][2] = yz - wx;
+
+    mtx[2][0] = xz - wy;
+    mtx[2][1] = yz + wx;
+    mtx[2][2] = 1 - xx - yy;
+
+    mtx[0][3] = 0.0f;
+    mtx[1][3] = 0.0f;
+    mtx[2][3] = 0.0f;
+}
+
+void Matrix34f::makeRT(const Vector3f &r, const Vector3f &t) {
+    const f32 sin[3] = {Mathf::sin(r.x), Mathf::sin(r.y), Mathf::sin(r.z)};
+    const f32 cos[3] = {Mathf::cos(r.x), Mathf::cos(r.y), Mathf::cos(r.z)};
+
+    const f32 c0_c2 = cos[0] * cos[2];
+    const f32 s0_s1 = sin[0] * sin[1];
+    const f32 c0_s2 = cos[0] * sin[2];
+
+    mtx[0][0] = (cos[1] * cos[2]);
+    mtx[1][0] = (cos[1] * sin[2]);
+    mtx[2][0] = (-sin[1]);
+
+    mtx[0][1] = (s0_s1 * cos[2]) - c0_s2;
+    mtx[1][1] = (s0_s1 * sin[2]) + c0_c2;
+    mtx[2][1] = (sin[0] * cos[1]);
+
+    mtx[0][2] = (c0_c2 * sin[1]) + (sin[0] * sin[2]);
+    mtx[1][2] = (c0_s2 * sin[1]) - (sin[0] * cos[2]);
+    mtx[2][2] = (cos[0] * cos[1]);
+
     mtx[0][3] = t.x;
-    mtx[1][0] = 2.0f * (q.v.x * q.v.y + q.w * q.v.y);
-    mtx[1][1] = 1.0 - 2.0f * (q.v.x * q.v.x + q.v.z * q.v.z);
-    mtx[1][2] = 2.0f * (q.v.y * q.v.z - q.w * q.v.x);
     mtx[1][3] = t.y;
-    mtx[2][0] = 2.0f * (q.v.x * q.v.z - q.w * q.v.y);
-    mtx[2][1] = 2.0f * (q.v.y * q.v.z + q.w * q.v.x);
-    mtx[2][2] = 1.0 - 2.0f * (q.v.x * q.v.x + q.v.y * q.v.y);
     mtx[2][3] = t.z;
 }
 
+void Matrix34f::makeR(const Vector3f &r) {
+    const f32 sin[3] = {Mathf::sin(r.x), Mathf::sin(r.y), Mathf::sin(r.z)};
+    const f32 cos[3] = {Mathf::cos(r.x), Mathf::cos(r.y), Mathf::cos(r.z)};
+
+    const f32 c0_c2 = cos[0] * cos[2];
+    const f32 s0_s1 = sin[0] * sin[1];
+    const f32 c0_s2 = cos[0] * sin[2];
+
+    mtx[0][0] = (cos[1] * cos[2]);
+    mtx[1][0] = (cos[1] * sin[2]);
+    mtx[2][0] = (-sin[1]);
+
+    mtx[0][1] = (s0_s1 * cos[2]) - c0_s2;
+    mtx[1][1] = (s0_s1 * sin[2]) + c0_c2;
+    mtx[2][1] = (sin[0] * cos[1]);
+
+    mtx[0][2] = (c0_c2 * sin[1]) + (sin[0] * sin[2]);
+    mtx[1][2] = (c0_s2 * sin[1]) - (sin[0] * cos[2]);
+    mtx[2][2] = (cos[0] * cos[1]);
+
+    mtx[0][3] = 0.0f;
+    mtx[1][3] = 0.0f;
+    mtx[2][3] = 0.0f;
+}
+
 void Matrix34f::makeZero() {
-    for (auto &n : a) {
-        n = 0.0f;
+    *this = Matrix34f::zero;
+}
+
+Matrix34f Matrix34f::multiplyTo(const Matrix34f &rhs) const {
+    Matrix34f mat;
+
+    mat(0, 0) = Mathf::fma(rhs(2, 0), mtx[0][2],
+            Mathf::fma(rhs(1, 0), mtx[0][1], rhs(0, 0) * mtx[0][0]));
+    mat(0, 1) = Mathf::fma(rhs(2, 1), mtx[0][2],
+            Mathf::fma(rhs(1, 1), mtx[0][1], rhs(0, 1) * mtx[0][0]));
```

Comment:
Is this maybe an instance where we'd be able to use `using Mathf` to make this dumpster fire of a function a little bit more readable?